### PR TITLE
JAMES-2233 WebAdmin responses more details when error

### PR DESCRIPTION
**Pull request by quynhn**

### CODE REVIEW

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/DomainsRoutes.java [hunk 1]
```diff
@@ -160,15 +162,16 @@ private void removeDomain(String domain) throws DomainListException {
...
             response.status(204);
         } catch (DomainListException e) {
-            LOGGER.info("{} already exists", request.params(DOMAIN_NAME));
-            response.status(204);
+            LOGGER.info("{} already exists", domainName);
+            throw halt(HttpStatus.NO_CONTENT_204, domainName + " already exists");
         } catch (IllegalArgumentException e) {
             LOGGER.info("Invalid request for domain creation");
-            response.status(400);
+            throw halt(HttpStatus.BAD_REQUEST_400, "Invalid request for domain creation " + domainName);
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/DomainsRoutes.java]: No cause?

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java [hunk 2]
```diff
@@ -158,13 +158,13 @@ public HaltException addToGroup(Request request, Response response) throws JsonE
 
     private void ensureRegisteredDomain(String domain) throws DomainListException {
         if (!domainList.containsDomain(domain)) {
-            throw halt(HttpStatus.FORBIDDEN_403);
+            throw halt(HttpStatus.FORBIDDEN_403, "server doesn't own the domain: " + domain);
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java]: Capital letter in the message?

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java [hunk 3]
```diff
@@ -158,13 +158,13 @@ public HaltException addToGroup(Request request, Response response) throws JsonE
...
         }
     }
 
     private void ensureNotShadowingAnotherAddress(MailAddress groupAddress) throws UsersRepositoryException {
         if (usersRepository.contains(groupAddress.asString())) {
-            throw halt(HttpStatus.CONFLICT_409);
+            throw halt(HttpStatus.CONFLICT_409, "requested group address is already used for another purpose");
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java]: Idem?

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java [hunk 4]
```diff
@@ -217,16 +217,16 @@ private MailAddress parseMailAddress(String address) {
             String decodedAddress = URLDecoder.decode(address, "UTF-8");
             return new MailAddress(decodedAddress);
         } catch (AddressException e) {
-            throw halt(HttpStatus.BAD_REQUEST_400);
+            throw halt(HttpStatus.BAD_REQUEST_400, "The group is not an address");
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java]: email address

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java [hunk 5]
```diff
@@ -217,16 +217,16 @@ private MailAddress parseMailAddress(String address) {
...
         } catch (UnsupportedEncodingException e) {
             LOGGER.error("UTF-8 should be a valid encoding");
-            throw halt(HttpStatus.INTERNAL_SERVER_ERROR_500);
+            throw halt(HttpStatus.INTERNAL_SERVER_ERROR_500, "Internal server error - Something went bad on the server side.");
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java]: No details?

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/UserRoutes.java [hunk 6]
```diff
@@ -135,12 +136,10 @@ private String removeUser(Request request, Response response) {
             response.status(204);
             return Constants.EMPTY_BODY;
         } catch (UsersRepositoryException e) {
-            response.status(204);
-            return "The user " + username + " does not exists";
+            throw halt(HttpStatus.NO_CONTENT_204, "The user " + username + " does not exists");
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/UserRoutes.java]: No need of message here

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/GlobalQuotaRoutes.java [hunk 7]
```diff
@@ -108,10 +111,10 @@ public void defineUpdateQuota() {
                 response.status(204);
             } catch (JsonExtractException e) {
                 LOGGER.info("Malformed JSON", e);
-                response.status(400);
+                throw halt(HttpStatus.BAD_REQUEST_400, "Malformed for input JSON: " + request.body());
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/GlobalQuotaRoutes.java]: Malformed JSON input

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/UserMailboxesRoutes.java [hunk 8]
```diff
@@ -127,13 +129,13 @@ public void defineDeleteUserMailbox() {
...
-                response.status(404);
+                throw halt(HttpStatus.NOT_FOUND_404, "Invalid delete on user mailboxes");
             } catch (MailboxHaveChildrenException e) {
                 LOGGER.info("Attempt to delete a mailbox with children");
-                response.status(409);
+                throw halt(HttpStatus.CONFLICT_409, "Attempt to delete a mailbox with children");
             } catch (IllegalArgumentException e) {
                 LOGGER.info("Attempt to create an invalid mailbox");
-                response.status(400);
+                throw halt(HttpStatus.BAD_REQUEST_400, "Attempt to create an invalid mailbox");
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/UserMailboxesRoutes.java]: Cause?

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/UserMailboxesRoutes.java [hunk 9]
```diff
@@ -127,13 +129,13 @@ public void defineDeleteUserMailbox() {
                 response.status(204);
             } catch (IllegalStateException e) {
                 LOGGER.info("Invalid delete on user mailbox", e);
-                response.status(404);
+                throw halt(HttpStatus.NOT_FOUND_404, "Invalid delete on user mailboxes");
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/UserMailboxesRoutes.java]: Cause?

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/UserMailboxesRoutes.java [hunk 10]
```diff
@@ -157,7 +159,7 @@ public void defineDeleteUserMailboxes() {
                 response.status(204);
             } catch (IllegalStateException e) {
                 LOGGER.info("Invalid delete on user mailboxes", e);
-                response.status(404);
+                throw halt(HttpStatus.NOT_FOUND_404, "Invalid delete on user mailboxes");
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/UserMailboxesRoutes.java]: Cause?

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/UserMailboxesRoutes.java [hunk 11]
```diff
@@ -183,14 +185,14 @@ public void defineMailboxExists() {
                 if (userMailboxesService.testMailboxExists(request.params(USER_NAME), new MailboxName(request.params(MAILBOX_NAME)))) {
                     response.status(204);
                 } else {
-                    response.status(404);
+                    throw halt(HttpStatus.NOT_FOUND_404, "Invalid get on user mailbox");
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/UserMailboxesRoutes.java]: 404 is IMO self explanatory

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/UserMailboxesRoutes.java [hunk 12]
```diff
@@ -100,8 +103,7 @@ public void defineGetUserMailboxes() {
                 return userMailboxesService.listMailboxes(request.params(USER_NAME));
             } catch (IllegalStateException e) {
                 LOGGER.info("Invalid get on user mailboxes", e);
-                response.status(404);
-                return Constants.EMPTY_BODY;
+                throw halt(HttpStatus.NOT_FOUND_404, "Invalid get on user mailboxes");
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/UserMailboxesRoutes.java]: 404 is IMO self explanatory

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/UserMailboxesRoutes.java [hunk 13]
```diff
@@ -217,10 +219,10 @@ public void defineCreateUserMailbox() {
                 response.status(204);
             } catch (IllegalStateException e) {
                 LOGGER.info("Invalid put on user mailbox", e);
-                response.status(404);
+                throw halt(HttpStatus.NOT_FOUND_404, "Invalid put on user mailbox");
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/UserMailboxesRoutes.java]: cause?

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/UserMailboxesRoutes.java [hunk 14]
```diff
@@ -217,10 +219,10 @@ public void defineCreateUserMailbox() {
...
             } catch (IllegalArgumentException e) {
                 LOGGER.info("Attempt to create an invalid mailbox");
-                response.status(400);
+                throw halt(HttpStatus.BAD_REQUEST_400, "Attempt to create an invalid mailbox");
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/UserMailboxesRoutes.java]: cause?

File: server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/routes/CassandraMigrationRoutes.java [hunk 15]
```diff
@@ -72,15 +72,13 @@ public void define(Service service) {
                 response.status(NO_CONTENT);
             } catch (NullPointerException | IllegalArgumentException e) {
                 LOGGER.info("Invalid request for version upgrade");
-                response.status(INVALID_VERSION);
+                throw halt(HttpStatus.BAD_REQUEST_400, "Invalid request for version upgrade " + e.getMessage());
```
Comment (heroandtn3) [inline server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/routes/CassandraMigrationRoutes.java]: never send the exception's message to the client because it could be a security risk, instead you can log it for later debugging

Comment (chibenwa) [reply server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/routes/CassandraMigrationRoutes.java]: The client here is necessarily an admin, with a **master, kill the world JWT**. This makes your assumption a bit weaker, IMO.

I would value a more explicit API, and I believe such security issues are not applicable.

Remember that your implementation strategy is to proxy webadmin for normal users, thus Admin OpenPaaS modules can filter that informations out.

I believe your **user <-> server** framework could be upgraded to **inter service communication**, and the considerations are not the same.

Food for thoughts.

**Comment by chibenwa** [comment]
After discussing with @heroandtn3 I changed the error format to:

```
{
    "error": {
       "statusCode": 400,
       "type": "invalidArgument",
        "explanation": "Invalid mail address for a group",
        "cause": "Invalid character '%' in position 5"
    }
}
```
```

File: server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/routes/CassandraMigrationRoutes.java [hunk 16]
```diff
@@ -72,15 +72,28 @@ public void define(Service service) {
...
+                    .haltError();
             } catch (MigrationException e) {
                 LOGGER.error("An error lead to partial migration process", e);
-                response.status(INTERNAL_ERROR);
-                response.body(e.getMessage());
+                throw ErrorResponder.builder()
+                    .statusCode(HttpStatus.INTERNAL_SERVER_ERROR_500)
+                    .type(ErrorType.SERVER_ERROR)
+                    .message("An error lead to partial migration process")
+                    .cause(e.getMessage())
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/routes/CassandraMigrationRoutes.java]: We should pass: 

```
    .cause(e)
```

This is more flexible and might enable having stacktrace....

File: server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/utils/ErrorResponder.java [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+    public HaltException haltError() {
+        try {
+            return halt(statusCode, new JsonTransformer().render(new ErrorDetail(statusCode, type.getType(), message, cause)));
+        } catch (JsonProcessingException e) {
+            return halt(statusCode);
+        }
+    }
+
+    private static class ErrorDetail {
+        private final int statusCode;
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/utils/ErrorResponder.java]: We might require some preconfitions, like status, type and message...

File: server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/utils/ErrorResponder.java [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+        this.message = message;
+        return this;
+    }
+
+    public ErrorResponder cause(String cause) {
+        this.cause = cause;
+        return this;
+    }
+
+    public HaltException haltError() {
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/utils/ErrorResponder.java]: No tests?

I would like to have assertions on the exception messages and ty to play with the various edge cases...

File: server/protocols/webadmin/webadmin-core/src/test/java/org/apache/james/webadmin/utils/ErrorResponderTest.java [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    @Test
+    public void haltErrorShouldReturnBodyWithStatusCodeWhenSetting() throws Exception {
+        assertThatThrownBy(() -> ErrorResponder.builder()
+                .statusCode(400)
+                .haltError())
+            .isInstanceOf(HaltException.class)
+            .matches(e -> hasStatus(e, 400))
+            .matches(e -> bodyHasErrorDetail(e, new ErrorResponder.ErrorDetail(400, "", "", "")));
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-core/src/test/java/org/apache/james/webadmin/utils/ErrorResponderTest.java]: I think **type** and **message** should be compulsory...

Comment (chibenwa) [reply server/protocols/webadmin/webadmin-core/src/test/java/org/apache/james/webadmin/utils/ErrorResponderTest.java]: Here **cause** should be and empty optional (be sure to have loaded Java 8 jackson bindings...)

File: server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/routes/CassandraMigrationRoutes.java (hunk 16 above)
Comment (strokyl) [reply server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/routes/CassandraMigrationRoutes.java]: Do we really want to show anyone the all stacktrace ?

File: server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/utils/ErrorResponder.java [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.webadmin.utils;
+
+import static spark.Spark.halt;
```
Comment (strokyl) [inline server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/utils/ErrorResponder.java]: Spark really ?

Comment (chibenwa) [reply server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/utils/ErrorResponder.java]: That"s the name of the java lib webadmin relly on. What's the issue?

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/DomainsRoutesTest.java [hunk 21]
```diff
@@ -157,10 +160,20 @@ public void getDomainsShouldDisplayAddedDomains() {
...
+                .contentType(ContentType.JSON)
+                .extract()
+                .body()
+                .jsonPath()
+                .getMap(".");
+
+            assertThat(errors)
+                .containsEntry("statusCode", 400)
+                .containsEntry("type", "InvalidArgument")
+                .containsEntry("message", "Invalid request for domain creation domain@domain");
```
Comment (strokyl) [inline server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/DomainsRoutesTest.java]: It is sad that contrary to the name of this test "putShouldReturnUserErrorWhenNameContainsAT" the error message does not let someone easy understand that the request fail because of a ``@`` in the username

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/DomainsRoutesTest.java [hunk 22]
```diff
@@ -173,12 +186,21 @@ public void putShouldReturnUserErrorWhenNameContainsUrlSeparator() {
...
+                .statusCode(400)
+                .contentType(ContentType.JSON)
+                .extract()
+                .body()
+                .jsonPath()
+                .getMap(".");
+
+            assertThat(errors)
+                .containsEntry("statusCode", 400)
+                .containsEntry("type", "InvalidArgument");
```
Comment (strokyl) [inline server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/DomainsRoutesTest.java]: Same here

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/GroupsRoutesTest.java [hunk 23]
```diff
@@ -370,74 +401,133 @@ public void setUp() throws Exception {
...
+                .extract()
+                .body()
+                .jsonPath()
+                .getMap(".");
+
+            assertThat(errors)
+                .containsEntry("statusCode", 400)
+                .containsEntry("type", "InvalidArgument")
+                .containsEntry("message", "The group is not an email address")
+                .containsEntry("cause", "Out of data at position 1 in 'not-an-address'");
```
Comment (strokyl) [inline server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/GroupsRoutesTest.java]: What does "Out of data at position 1 in 'not-an-address'" mean ?

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/GroupsRoutesTest.java [hunk 24]
```diff
@@ -370,74 +401,133 @@ public void setUp() throws Exception {
...
         }
 
         @Test
         public void deleteRequiresTwoPathParams() {
             when()
                 .delete(GROUP1)
             .then()
-                .statusCode(HttpStatus.BAD_REQUEST_400);
+                .statusCode(HttpStatus.BAD_REQUEST_400)
+                .body(is(""));
```
Comment (strokyl) [inline server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/GroupsRoutesTest.java]: No message error here? Why ?

File: src/site/markdown/server/manage-webadmin.md [hunk 25]
```diff
@@ -7,6 +7,14 @@ The web administration supports for now the CRUD operations on the domains,the u
 
 Please also note **webadmin** is only enabled with **Guice**. You can not use it when using James with **Spring**, as the required injections are not implemented.
 
+In case of any error, the system will return an error message which is json format like this:
+{
+    statusCode: <error_code>,
+    type: <error_type>,
+    message: <the_error_message>
+    cause: <the_detail_message_from_throwable>
```
Comment (strokyl) [inline src/site/markdown/server/manage-webadmin.md]: Why _ instead of space. I think space are fine and more human friendly to read

File: server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/utils/ErrorResponder.java (hunk 20 above)
Comment (strokyl) [reply server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/utils/ErrorResponder.java]: No I just think it was weird is it related to "spark hadoop" ?
I have no issue with it

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/DomainsRoutesTest.java (hunk 21 above)
Comment (quynhn) [reply server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/DomainsRoutesTest.java]: I would like to keep all the function and test function as they are. I just add more assertion error information in case of error.

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/DomainsRoutesTest.java (hunk 22 above)
Comment (quynhn) [reply server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/DomainsRoutesTest.java]: idem

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/GroupsRoutesTest.java (hunk 23 above)
Comment (quynhn) [reply server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/GroupsRoutesTest.java]: That is the message from throwable, I don't want to refactor it

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/GroupsRoutesTest.java (hunk 24 above)
Comment (quynhn) [reply server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/GroupsRoutesTest.java]: that is the purpose of API, no message when halt

File: src/site/markdown/server/manage-webadmin.md (hunk 25 above)
Comment (quynhn) [reply src/site/markdown/server/manage-webadmin.md]: I tried but "space" looks strange in case of variable. Can we live with this?

File: server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/routes/CassandraMigrationRoutes.java [hunk 26]
```diff
@@ -72,15 +72,28 @@ public void define(Service service) {
                 response.status(NO_CONTENT);
             } catch (NullPointerException | IllegalArgumentException e) {
                 LOGGER.info("Invalid request for version upgrade");
-                response.status(INVALID_VERSION);
+                throw ErrorResponder.builder()
+                    .statusCode(HttpStatus.BAD_REQUEST_400)
+                    .type(ErrorType.INVALID_ARGUMENT)
+                    .message("Invalid request for version upgrade")
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/routes/CassandraMigrationRoutes.java]: it duplicates every log sentences, not a good idea

File: server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java [hunk 27]
```diff
@@ -145,12 +145,23 @@ public void postShouldReturnConflictWhenMigrationOnRunning() throws Exception {
...
+            .statusCode(400)
+            .contentType(ContentType.JSON)
+            .extract()
+            .body()
+            .jsonPath()
+            .getMap(".");
+
+        assertThat(errors)
+            .containsEntry("statusCode", 400)
+            .containsEntry("type", "InvalidArgument")
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java]: it's already what 400 is saying : Bad Input

File: server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java [hunk 28]
```diff
@@ -145,12 +145,23 @@ public void postShouldReturnConflictWhenMigrationOnRunning() throws Exception {
...
-            .statusCode(400);
+            .statusCode(400)
+            .contentType(ContentType.JSON)
+            .extract()
+            .body()
+            .jsonPath()
+            .getMap(".");
+
+        assertThat(errors)
+            .containsEntry("statusCode", 400)
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java]: duplicates HTTP code, not a good idea IMO

File: server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java [hunk 29]
```diff
@@ -145,12 +145,23 @@ public void postShouldReturnConflictWhenMigrationOnRunning() throws Exception {
...
+            .containsEntry("type", "InvalidArgument")
+            .containsEntry("message", "Invalid request for version upgrade")
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java]: does it add anything we don't know already ?

File: server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java [hunk 30]
```diff
@@ -145,12 +145,23 @@ public void postShouldReturnConflictWhenMigrationOnRunning() throws Exception {
...
+            .containsEntry("cause", "For input string: \"NonInt\"");
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java]: `cause : For input string`
doesn't look like a cause ...

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/DomainsRoutesTest.java (hunk 21 above)
Comment (strokyl) [reply server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/DomainsRoutesTest.java]: I get you point of view but if it is easy to fix, it is the good opportunity to improve James.
Clear error message are really important I think for the UX

File: src/site/markdown/server/manage-webadmin.md (hunk 25 above)
Comment (strokyl) [reply src/site/markdown/server/manage-webadmin.md]: Ok

File: server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java [hunk 31]
```diff
@@ -175,12 +186,23 @@ public void postShouldDoMigrationToNewVersion() throws Exception {
...
+            .extract()
+            .body()
+            .jsonPath()
+            .getMap(".");
+
+        assertThat(errors)
+            .containsEntry("statusCode", 410)
+            .containsEntry("type", "WrongState")
+            .containsEntry("message", "The migration requested can not be performed")
+            .containsEntry("cause", "Current version is already up to date");
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java]: are you sure ? then why it says "gone" ?

File: server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java (hunk 29 above)
Comment (strokyl) [reply server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java]: It just potentially explain to a user what ``400`` mean. It is redondant but armless I will let it for the moment.
@quynhn what do you do ?

File: server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java (hunk 30 above)
Comment (strokyl) [reply server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java]: This string come from a "e.getMessage()" and the "e" in question come from a json parse.
I don't really see what I can do to make that better sorry

File: server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java (hunk 31 above)
Comment (strokyl) [reply server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java]: Good question

Comment (strokyl) [reply server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java]: Hum I think the logic behind that it was that this request was available upon a time.
Because, in rest everything is a resource, this request is represented by a ressource that is not here anymore.
But I think BAD_REQUEST will be more appropriate I am going to change for that

File: server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/utils/ErrorResponder.java (hunk 20 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/utils/ErrorResponder.java]: Not at all...

http://sparkjava.com/

File: server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java (hunk 30 above)
Comment (mbaechler) [reply server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java]: You described the problem : a user facing message generated by a component we don't manage
Looks like a bad idea

File: server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java (hunk 31 above)
Comment (mbaechler) [reply server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java]: the problem is that upgrade is not a resource.

**Comment by chibenwa** [comment]
Merged
